common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.containerpool
+
+import akka.actor.ActorSystem
+import scala.collection.mutable.HashMap
+import scala.collection.mutable.MultiMap
+import scala.collection.mutable.Set
+import scala.concurrent.Future
+import whisk.common.Logging
+import whisk.common.TransactionId
+import whisk.core.WhiskConfig
+import whisk.core.entity.ByteSize
+import whisk.core.entity.ExecManifest
+import whisk.core.entity.InstanceId
+import whisk.spi.Spi
+
+/**
+ * An abstraction for Container creation
+ */
+trait ContainerFactory {
+
+  /** create a new Container */
+  def createContainer(tid: TransactionId,
+                      name: String,
+                      actionImage: ExecManifest.ImageName,
+                      userProvidedImage: Boolean,
+                      memory: ByteSize)(implicit config: WhiskConfig, logging: Logging): Future[Container]
+
+  /** perform any initialization */
+  def init(): Unit
+
+  /** cleanup any remaining Containers; should block until complete; should ONLY be run at startup/shutdown */
+  def cleanup(): Unit
+}
+object ContainerFactory {
+
+  /** build the parameters for docker run */
+  def dockerRunParameters(config: WhiskConfig): MultiMap[String, String] = {
+    val dockerRunParameters = new HashMap[String, Set[String]] with MultiMap[String, String]
+    dockerRunParameters.addBinding("--cap-drop", "NET_RAW")
+    dockerRunParameters.addBinding("--cap-drop", "NET_ADMIN")
+    dockerRunParameters.addBinding("--ulimit", "nofile=1024:1024")
+    dockerRunParameters.addBinding("--pids-limit", "1024")
+    val dnsServers = config.invokerContainerDns
+    dnsServers.map(d => dockerRunParameters.addBinding("--dns", d))
+    dockerRunParameters
+  }
+}
```

Comment:
Not sure if using a `Map` here is useful. For additional arguments I found it mostly useful to just use a flat `Seq` and pass stuff that way.

Alternatively we could have a type encapsulating all the options on a docker run and propagate it that way. But please do not use a mutable map here. If you think a Map is the right thing to do, let's at least use an immutable one (I know you can't use MultiMap then)

Reply:
At least partly I used Map because a) all the options I think that should be settable here are key-value and b) I happen to know in mesos we will have to translate this into a Map :(

I think the options will boil down to either using Seq (which at least some alternative container factories will have to decode to a Map) or writing an immutable version of MultiMap. I'm not wild about either of these prospects, let me know if you have other ideas.

Reply:
As seen in the method signature, `MultiMap` is just some sugar over `Map[String, Set[String]]`, so you can translate your mutable code to:

```scala
Map(
  "--cap-drop" -> Set("NET_RAW", "NET_ADMIN"),
  "--ulimit" -> Set("nofile=1024:1024"),
  "--pids-limit" -> Set("1024"),
  "--dns" -> config.invokerContainerDns.toSet
)
```

Which I'd prefer to keep the interface immutable. MultiMap-sugar is arguably not helping much here anyway?

Reply:
Also: Shouldn't this method exist only for the docker implementation?

Reply:
Agree on immutable MultiMap, will make those changes.

Run parameters are possible in other non-docker-cli container management systems (e.g. mesos, guessing kube as well), so it should be useable by any SPI impl.

Reply:
Still the method is called `dockerRunParameters` but it's part of the general `ContainerFactory`. It's not part of any interface, so it should live in `DockerContainerFactory`?

Reply:
Are you suggesting to rename it to `containerRunParameters`? or that there may be some non-docker container support, like lxc/oci/etc? 

The problem with "moving it to DockerContainerFactory" is that DockerContainerFactory shouldn't be used if some other container factory is used, but these parameters are still important. 

This may be confusing terminology, but currently the notion of "container" exists only in the form of "docker container", but the ContainerFactory is allowing multiple approaches to creating "docker containers" (as opposed to creating "other non-docker containers"). Its a difference of using docker cli to create a docker container, vs using some other system (which in turn uses docker cli) to create a docker container.